policy/src/main/java/org/apache/brooklyn/policy/ha/PropagatePrimaryEnricher.java

```diff
@@ -0,0 +1,204 @@
+        blacklistedSensors.addAll(((EntityInternal)entity).getMutableEntityType().getSensors().keySet());
+        for (Sensor<?> s: Propagator.SENSORS_NOT_USUALLY_PROPAGATED) {
+            blacklistedSensors.add(s.getName());
+        }
+        blacklistedSensors.addAll(MutableSet.of(getConfig(PRIMARY_SENSOR_NAME), getConfig(ElectPrimaryConfig.PRIMARY_WEIGHT_NAME)));
+        
+        subscriptions().subscribe(entity, Sensors.newSensor(Entity.class, config().get(PRIMARY_SENSOR_NAME)), this);
+        onEvent(null);
+    }
+
+    @Override
+    public synchronized void onEvent(SensorEvent event) {
+        Entity primary = entity.getAttribute( Sensors.newSensor(Entity.class, config().get(PRIMARY_SENSOR_NAME)) );
+        if (!Objects.equal(primary, lastPrimary)) {
+            log.debug("Removing propagated items from "+lastPrimary+" at "+entity);
+            
+            // remove propagator
+            if (propagator!=null) {
+                entity.enrichers().remove(propagator);
+                propagator = null;
+            }
+
+            // remove propagated effectors
+            if (effectorsAddedForPrimary!=null) {
+                log.debug("Removing propagated effectors from "+lastPrimary+" at "+entity+": "+effectorsAddedForPrimary);
+                for (String effN: effectorsAddedForPrimary) {
+                    Effector<?> effE = ((EntityInternal)entity).getMutableEntityType().getEffector(effN);
+                    if (effE!=null) {
+                        ((EntityInternal)entity).getMutableEntityType().removeEffector(effE);
+                    }
+                }
+                effectorsAddedForPrimary = null;
+            }
+            
+            // remove all but blacklisted sensors
+            Set<AttributeSensor<?>> sensorsToRemove = MutableSet.of();
+            for (AttributeSensor<?> s: ((SensorSupportInternal)entity.sensors()).getAll().keySet()) {
+                if (!blacklistedSensors.contains(s.getName())) {
+                    sensorsToRemove.add(s);
+                }
+            }
+            if (!sensorsToRemove.isEmpty()) {
+                log.debug("Removing propagated sensors from "+lastPrimary+" at "+entity+": "+sensorsToRemove);
+                for (AttributeSensor<?> s: sensorsToRemove) {
+                    ((SensorSupportInternal)entity.sensors()).remove(s);
+                }
+            }
+            
+            
+            if (primary!=null) {
+                // add effectors
+                propagateEffectors(primary);
+                
+                // add propagator
+                addPropagatorEnricher(primary);
+            }
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    protected void addPropagatorEnricher(Entity primary) {
+        EnricherSpec<Propagator> spec = EnricherSpec.create(Propagator.class);
+        if (Boolean.TRUE.equals( getConfig(PROPAGATING_ALL) )) {
+            spec.configure(PROPAGATING_ALL, true);
```

Comment:
we need special logic for `ALL_BUT` to add extra exceptions, but otherwise i liked this until i discovered it actually looks like:

    spec.configureIfNotNull((ConfigKey<Object>)(ConfigKey<?>)PROPAGATING, config().getRaw(PROPAGATING));

it does mean all config gets evaluated once only here which is a bit of a limitation.  but we need to do that anyway for `ALL_BUT` so on balance i'm tempted to leave it as is.